### PR TITLE
Update Python bindings to build with recent awkward-array and improve testing

### PR DESCRIPTION
**Pull request by gillins**
Apologies for the size of this PR. 

It attempts to make the Python bindings (mainly for access to the 'neighbours' functionality) more usable with the following changes:

1. Actually compile successfully against recent versions of awkward-array. Thanks to @agoose77 and @jpivarski for their kind help here. Awkward arrays are used for reading and writing an extra field in the Raster Attribute Table which is the (spatially) neighbouring rows for each class. This is mainly used for processing outputs of a segmentation, which is a common use case for KEA - and ideal for awkward arrays since they can vary in size enormously. GDAL does not have access to this (and fair enough, it's a KEA format specific feature). Also in this PR as a kind of experiment I've added reading and writing of rows of string columns as awkward arrays. GDAL does this as 2D arrays of strings which is kind of clunky. I've also increased the use of `pybind11` (as used in the awkward-array examples) to make things consistent.
2. Add some other KEA specific features: a) creating boolean columns and reading and writing to them as bools - GDAL does not support creation of this column type but can read and write them as ints from/to a KEA file and b) access int columns as int64's - which is how they are actually stored in KEA - GDAL only allows access to them as int32's.
3. De-couple these Python bindings from GDAL somewhat. Because I assume that most users of this functionality will already have the file open as a GDAL dataset, I accept this as the handle to operate on. Otherwise I'd be forced to reproduce all the functionality GDAL already provides which seems unnecessary. However this creates a circular dependency with GDAL when building. I've worked around this by only linking to GDAL at runtime and faking enough of the GDAL functionality for testing. See the comments at the top of `fakegdal.cpp`. I think this is an important step and will allow building of these bindings on conda-forge which wasn't possible before.
4. As a result of not having access to full GDAL (see above) add enough functionality for building the neighbours and running a test suite without GDAL (ie be able to just use `extrat` module for this).

I envisage this module as kind of a testing ground for KEA specific features. Some things may be moved into GDAL, or otherwise changed - just be warned there could be some churn before we settle what this will look like. 

cc: @neilflood @petebunting @bhjolly @shepjd @petescarth

### CODE REVIEW

File: python/CMakeLists.txt [hunk 1]
```diff
@@ -28,32 +29,67 @@ else()
...
-message(STATUS "Libraries: ${CPU-KERNELS} ${LIBAWKWARD} ${LIBKEA} ${GDAL_LIBRARIES}")
+
+find_package(HDF5 COMPONENTS CXX HL REQUIRED)
 
 pybind11_add_module(extrat extrat.cpp)
+target_include_directories(extrat BEFORE PRIVATE "${AWKWARD_INCLUDE1}" PRIVATE "${AWKWARD_INCLUDE2}" PRIVATE "${AWKWARD_INCLUDE3}" PRIVATE "${KEALIB_INCLUDE}" PRIVATE "${Python_NumPy_INCLUDE_DIRS}")
 target_compile_definitions(extrat PRIVATE NPY_NO_DEPRECATED_API=NPY_1_7_API_VERSION)
 set_target_properties(extrat PROPERTIES CXX_VISIBILITY_PRESET hidden)
-target_link_libraries(extrat PRIVATE ${CPU-KERNELS} ${LIBAWKWARD} ${LIBKEA} ${GDAL_LIBRARIES})
+target_link_libraries(extrat PRIVATE ${LIBKEA})
```
[the remark this answers is not in the document]
Comment (gillins) [reply python/CMakeLists.txt]: Thanks @agoose77 ! I think I've managed to work this out - `pybind11` can be a bit obscure....

File: python/CMakeLists.txt [hunk 2]
```diff
@@ -28,32 +29,67 @@ else()
...
+)
+# don't use FetchContent_MakeAvailable(awkward) - it will install - see https://stackoverflow.com/questions/65527126/disable-install-for-fetchcontent
+FetchContent_GetProperties(awkward)
+if(NOT awkward_POPULATED)
+  FetchContent_Populate(awkward)
+  add_subdirectory(${awkward_SOURCE_DIR} ${awkward_BINARY_DIR} EXCLUDE_FROM_ALL)
+endif()
+set(AWKWARD_INCLUDE1 "${awkward_SOURCE_DIR}/layout-builder")
+set(AWKWARD_INCLUDE2 "${awkward_SOURCE_DIR}/builder-options")
+set(AWKWARD_INCLUDE3 "${awkward_SOURCE_DIR}/growable-buffer")
```
[the remark this answers is not in the document]
Comment (gillins) [reply python/CMakeLists.txt]: Done.

File: python/CMakeLists.txt [hunk 3]
```diff
@@ -28,32 +29,67 @@ else()
...
 
-add_subdirectory(pybind11)
+find_package(pybind11 CONFIG REQUIRED)
 
-find_library(CPU-KERNELS awkward-cpu-kernels REQUIRED HINTS ${AWKWARD_LIBRARIES})
-find_library(LIBAWKWARD awkward REQUIRED HINTS ${AWKWARD_LIBRARIES})
 find_library(LIBKEA ${LIBKEA_NAME} REQUIRED HINTS ${KEALIB_LIBRARIES})
-message(STATUS "Libraries: ${CPU-KERNELS} ${LIBAWKWARD} ${LIBKEA} ${GDAL_LIBRARIES}")
+
+find_package(HDF5 COMPONENTS CXX HL REQUIRED)
```
[the remark this answers is not in the document]
Comment (gillins) [reply python/CMakeLists.txt:60]: Hmm probably... Any pointers on how to do this? I'll add this in another PR.

[the remark this answers is not in the document]
Comment (agoose77) [reply python/CMakeLists.txt:60]: It's a bit of a faff, but there are good guides! I'll have a look and share them here :)

**Pull request merged**